src/libraries/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.Unescaping.cs

```diff
@@ -421,47 +523,120 @@ internal static void Unescape(ReadOnlySpan<byte> source, Span<byte> destination,
+                        bool success = rune.TryEncodeToUtf8(destination.Slice(written), out int bytesWritten);
 #else
-                        EncodeToUtf8Bytes((uint)scalar, destination.Slice(written), out int bytesWritten);
+                        bool success = TryEncodeToUtf8Bytes((uint)scalar, destination.Slice(written), out int bytesWritten);
 #endif
+                        if (!success)
+                        {
+                            goto DestinationTooShort;
+                        }
+
                         Debug.Assert(bytesWritten <= 4);
                         written += bytesWritten;
-                    }
+                        break;
                 }
-                else
+
+                if (++idx == source.Length)
                 {
-                    destination[written++] = currentByte;
+                    goto Success;
+                }
+
+                if (source[idx] != JsonConstants.BackSlash)
+                {
+                    ReadOnlySpan<byte> remaining = source.Slice(idx);
+                    int nextUnescapedSegmentLength = remaining.IndexOf(JsonConstants.BackSlash);
+                    if (nextUnescapedSegmentLength < 0)
+                    {
+                        nextUnescapedSegmentLength = remaining.Length;
+                    }
+
+                    if ((uint)(written + nextUnescapedSegmentLength) >= (uint)destination.Length)
+                    {
+                        goto DestinationTooShort;
+                    }
+
+                    Debug.Assert(nextUnescapedSegmentLength > 0);
+                    switch (nextUnescapedSegmentLength)
+                    {
+                        case 1:
+                            destination[written++] = source[idx++];
+                            break;
+                        case 2:
+                            destination[written++] = source[idx++];
+                            destination[written++] = source[idx++];
+                            break;
+                        case 3:
+                            destination[written++] = source[idx++];
+                            destination[written++] = source[idx++];
+                            destination[written++] = source[idx++];
+                            break;
+                        default:
+                            remaining.Slice(0, nextUnescapedSegmentLength).CopyTo(destination.Slice(written));
+                            written += nextUnescapedSegmentLength;
+                            idx += nextUnescapedSegmentLength;
+                            break;
+                    }
+
+                    Debug.Assert(idx == source.Length || source[idx] == JsonConstants.BackSlash);
+
+                    if (idx == source.Length)
+                    {
+                        goto Success;
+                    }
                 }
             }
+
+        Success:
+            return true;
+
+        DestinationTooShort:
+            return false;
```

Comment:
why not just `return false` directly? do we need to use `goto DestinationTooShort`? False return means `DestinationTooShort` anyway. If you think we might change that in the future perhaps consider:
`const bool DestinationTooShort = false;` and `return DestinationTooShort`

Reply:
I'd suggest filing issue on improving this rather than making code harder to read

Reply:
It's fairly pervasive in the hot code paths in both STJ and STEW, so we can look at it later.

Reply:
We should at least open the issue to track this work